rpc_sidecar/src/node_client.rs

```diff
@@ -305,151 +296,119 @@ impl Error {
-                    if !config.max_attempts.can_attempt(current_attempt) {
+                    if !config
+                        .exponential_backoff
+                        .max_attempts
+                        .can_attempt(current_attempt)
+                    {
                         anyhow::bail!(
                             "Couldn't connect to node {} after {} attempts",
-                            addr,
+                            config.address,
                             current_attempt - 1
                         );
                     }
                     tokio::time::sleep(Duration::from_millis(wait)).await;
-                    wait = (wait * config.coefficient).min(config.max_delay_ms);
+                    wait = (wait * config.exponential_backoff.coefficient)
+                        .min(config.exponential_backoff.max_delay_ms);
                 }
-            }
+            };
         }
     }
+
+    async fn reconnect(
+        config: &NodeClientConfig,
+    ) -> Result<Framed<TcpStream, BinaryMessageCodec>, AnyhowError> {
+        let disconnected_start = Instant::now();
+        inc_disconnect();
+        error!("node connection closed, will attempt to reconnect");
+        let stream = Self::connect_with_retries(config).await?;
+        info!("connection with the node has been re-established");
+        observe_reconnect_time(disconnected_start.elapsed());
+        Ok(stream)
+    }
 }
 
 #[async_trait]
-impl NodeClient for JulietNodeClient {
+impl NodeClient for FramedNodeClient {
     async fn send_request(&self, req: BinaryRequest) -> Result<BinaryResponseAndRequest, Error> {
-        let payload = encode_request(&req).expect("should always serialize a request");
-        let request_guard = self
-            .client
-            .read()
-            .await
-            .create_request(ChannelId::new(0))
-            .with_payload(payload.into())
-            .queue_for_sending()
-            .await;
-        let response = request_guard
-            .wait_for_response()
-            .await
-            .map_err(|err| Error::RequestFailed(err.to_string()))?
-            .ok_or(Error::NoResponseBody)?;
-        let resp = bytesrepr::deserialize_from_slice(&response)
+        let payload =
+            BinaryMessage::new(encode_request(&req).expect("should always serialize a request"));
+        let mut client = self.client.write().await;
+
+        if let Err(err) = tokio::time::timeout(
+            Duration::from_secs(self.config.message_timeout_secs),
+            client.send(payload),
+        )
+        .await
+        .map_err(|_| Error::RequestFailed("timeout".to_owned()))?
+        {
+            warn!(
+                addr = %self.config.address,
+                err = display_error(&err),
+                "binary port client handler error"
+            );
+            return Err(Error::RequestFailed(err.to_string()));
+        };
+
+        let Ok(maybe_response) = tokio::time::timeout(
```

Comment:
Similiar issue, in the case of a timeout, the response is not cleared :)

Reply:
Reconnection logic has been rewritten in 96e2a3ec5c696d2c98a49e01886b10668b888f5d